File: fall-2025/cecs-342/quick-sort/main.c
#include <stdio.h>

// AI explanation here:
/*
  Quick sort is a divide-and-conquer algorithm:
  1. Pick a pivot (an element from the array).
  2. Partition: Rearrange the array so:
  3. All elements less than the pivot are on the left.
  4. All elements greater than the pivot are on the right.
  5. Recurse: Apply the same steps to the left and right sub-arrays.
  6. Stop when sub-arrays have 0 or 1 element → they’re already sorted.
*/

// TOOLS USED:
// vscode, gcc, ghosty terminal, chat gpt for explaining algorithm

void quick_sort(int *arr, int size) {
  // arr [3, 24, 5, 23, 2, 19, 21]
  // size = 7

  if (size <= 1) {
    return;
  }

  int pivot = arr[size / 2];
  int i = 0;
  int j = size - 1; // start at the ned

  while (i <= j) {
    // move pointer "i" up if less than pivot
    while (arr[i] < pivot) {
      ++i;
    }

    // move pointer "j" down if greater than pivot
    while (arr[j] > pivot) {
      --j;
    }

    // perform swap
    if (i <= j) {
      int temp = arr[i];
      arr[i] = arr[j];
      arr[j] = temp;

      ++i;
      --j;
    }
  }

  // BY THIS POINT i IS GREATER THAN j
  // ofc on each recursive call
  // visual [3, 2, 5, 23, 24, 21, 19]
  //               j      i

  if (j > 0) {
    // opposite j acts as i
    // in other words we are working on the left hand side
    // this also handles the previous pivot point
    quick_sort(arr, j + 1);
  }

  if (i < size) {
    // left side gets removed 3, 2, 5, 23
    // visual [24, 21, 19]
    //          i
    // size - i === 7 - 4 = 3 (which is our new size)
    quick_sort(arr + i, size - i);
  }
}

void run_test_case(int *arr, int size) {
  printf("Input: {");
  for (int i = 0; i < size; i++) {
    // if not at end print with comma
    if (i + 1 < size) {
      printf("%d, ", arr[i]);
    } else {
      printf("%d", arr[i]);
    }
  }
  printf("}\n");

  quick_sort(arr, size);

  printf("Sorted: {");
  for (int i = 0; i < size; i++) {
    // if not at end print with comma
    if (i + 1 < size) {
      printf("%d, ", arr[i]);
    } else {
      printf("%d", arr[i]);
    }
  }
  printf("}\n\n");
}

int main(void) {
  printf("CECS 342 Assignment 1\n");
  printf("Name: Carlos Aguilera\n");
  printf("Algorithm: quick sort, Language: C\n\n");

  run_test_case(NULL, 0);

  int test_case_input_1[] = {3};
  run_test_case(test_case_input_1, 1);

  int test_case_input_2[] = {3, 2, 1};
  run_test_case(test_case_input_2, 3);

  int test_case_input_3[] = {42, -33, 102, -74, 30, 1, 1000, -756, 33, -7, 10};
  run_test_case(test_case_input_3, 11);

  return 0;
}

File: fall-2025/cecs-342/quick-sort/Main.hs
-- Main.hs
--
-- CECS 342 Assignment 1
-- Quick sort in Haskell

-- TOOLS USED:
-- vscode, gcc, ghosty terminal, chat gpt for explaining algorithm

module Main where

-- Quick sort a list.
-- qsort :: Ord a => [a] -> [a]
-- qsort xs
--   | null xs        = []          -- size < 1
--   | length xs == 1 = xs          -- size == 1
--   | otherwise =
--       let i = 0
--           j = length xs - 1
--       in helper xs i j
--   where
--     -- Local helper: put the *type signature* and *definition* here.
--     pivot :: Int
--     pivot = xs[length xs / 2]

--     helper :: Ord a => [a] -> Int -> Int -> null
--     helper arr i j
--       | i > j        = null          -- size < 1
--       | otherwise =
--         if arr[i] < pivot
--           helper arr i + 1 j
        
--         if arr[j] > pivot
--           helper arr i j - 1
        
--         if j > 0
--           qsort xs j + 1
        
--         if i < length xs
--           qsort (xs + i) (length xs - i)

-- I did get some help from gpt to try and make my idea above come to life
-- But overall logic was me, I also comented each haskell line to ensure I understand it
qsort :: Ord a => [a] -> [a]
qsort xs
  | n <= 1    = xs
  | otherwise =
      let mid   = n `div` 2 -- n / 2
          pivot = xs !! mid -- super weird but essentially is xs[mid]
          rest  = take mid xs ++ drop (mid+1) xs -- remove mid and concat left and right
          left  = filter (<= pivot) rest -- filter any element thats <= pivot
          right = filter (> pivot) rest -- filter any element that is > pivot
      in qsort left ++ [pivot] ++ qsort right -- concat back together with the now sorted array and include pivot
  where
    n = length xs


-- Call a sort function and print input and output.
testSort sort input = do
  putStr "Input:  "
  print input
  putStr "Sorted: "
  print (sort input)

-- Call some tests.
main = do
  putStrLn "CECS 342 Assignment 1"
  putStrLn "Name: Carlos Aguilera"
  putStrLn "Algorithm: quick sort, Language: Haskell\n"
  
  testSort qsort ([]::[Int])
  testSort qsort [1]
  testSort qsort [3, 2, 1]
  testSort qsort [102, -43, 5, 645, 4, 34, -6, 776, 445, 45]
  testSort qsort ["the",  "quick",  "brown", "fox",
                  "jumps", "over", "the", "lazy", "dog"]
